### PR TITLE
Sphinx generated documentation

### LINKED ISSUE #465: Automatic docs
I know it was proposed before and I tabled it but I think it's time to reconsider so docs can scale easier.

Let's use whatever PyTorch uses and do documentation in the code (i assume that's how it's done?).

Anyone want to take a stab at this? @Borda

### PR DESCRIPTION
**Pull request by Borda**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/williamFalcon/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #465.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## The main changes:
- create Sphinx documentation
- move docs from MD to docstring in particular package/modules
- CI: add building documentation
- rename `root_module.root_module` -> `base_module.lightning`

## Watch
```
cd docs
pip install -r requirements.txt
make clean
make html
make latexpdf
```
then open `pytorch-lightning/docs/build/html/index.html`

### CODE REVIEW

**Comment by Borda** [comment]
@lr1d do you have practical experience with customization of `pytorch_sphinx_theme`?
I was able to add it and build but I have not found an option how to overwrite short links and navigation elements...

**Comment by Ir1d** [comment]
@Borda I didn't modify the theme.. perhaps we need to maintain a fork by modifying `pytorch_sphinx_theme` itself

**Comment by Borda** [comment]
@williamFalcon @Ir1d @neggert @jeffling could you pls have look and new docs...
you can build it locally or I can issue a public version on ReadTheDocs (but it will be public) :]

**Comment by Borda** [comment]
also @williamFalcon I would like to know what paths shall we use for doc/tutorial/blog/etc.

**Comment by neggert** [comment]
Got everything checked out and building. Lots of warnings, but everything seems to have worked okay. 

First off, thanks for tackling this. This is a huge undertaking and it's going to be a great improvement to the package.

A couple notes:

* Template needs to be tweaked a little bit to remove or fix dead links in the header. 
* Need to replace they PyTorch logo with something else (do we have an official logo for Lightning?)
* Because a lot of the trainer is split up into mixins, the auto-generated doc tree is a little bit confusing. It might be more effective to do what [PyTorch does](https://github.com/pytorch/pytorch/tree/master/docs/source): Create the high-level structure of the docs manually, but pull in autodoc for individual classes/functions/methods. That way we can organize things in a way that makes more sense for the user.

I don't want the perfect to be the enemy of the good, though. A lot of work has gone into getting this all set up, so maybe we just merge this with the understanding that it needs a bit more tweaking before we set RTD to public.

**Comment by williamFalcon** [comment]
@Borda this looks amazing! i agree with @neggert, let’s merge and fix to get it to a spot to publicly launch.

**Comment by Borda** [comment]
> Got everything checked out and building. Lots of warnings, but everything seems to have worked okay.

I saw them too, but I wanted to get a feedback if this is the right way to go before I will invest more time to it...

> * Template needs to be tweaked a little bit to remove or fix dead links in the header.

Sure, depending on how large changes will we need we may need to fork the theme, but first I would need to know what links to keep and what to drop... @williamFalcon 

> * Need to replace they PyTorch logo with something else (do we have an official logo for Lightning?)

yes, there is but it is the sam eas icon of an epic task in JIRA :)

> * Because a lot of the trainer is split up into mixins, the auto-generated doc tree is a little bit confusing. It might be more effective to do what [PyTorch does](https://github.com/pytorch/pytorch/tree/master/docs/source): Create the high-level structure of the docs manually, but pull in autodoc for individual classes/functions/methods. That way we can organize things in a way that makes more sense for the user.

I will have a look

> I don't want the perfect to be the enemy of the good, though. A lot of work has gone into getting this all set up, so maybe we just merge this with the understanding that it needs a bit more tweaking before we set RTD to the public.

I agree that we (me) shall get to stage that it is publishable, but then we shall merge it or move it to created `docs` branch so also others can help because this just reveals lost of issues which were simply hidden behind the static pages...

**Comment by williamFalcon** [comment]
the lightning logo is on the readme :)

**Comment by Borda** [comment]
image
https://bestin-it.com/jira-lesson-08-board-workflow-configuration/

**Comment by awaelchli** [comment]
@Borda I see no problem. Lightning IS epic. ;)

**Comment by Borda** [comment]
@neggert I have forked the PyTorch theme and did some adjustment but somehow the logo is not visible, it is probably something in CSS which I do not understand... Could you have look?

**Comment by Ir1d** [comment]
Have you tried replacing this file directly? https://github.com/Borda/lightning_sphinx_theme/blob/master/images/logo.svg

**Comment by Borda** [comment]
> Have you tried replacing this file directly? https://github.com/Borda/lightning_sphinx_theme/blob/master/images/logo.svg

these images are irrelevant, you need to look at these inside the package...

**Comment by Borda** [comment]
@williamFalcon @Ir1d @neggert @jeffling I would go for a merge, so pls could you do a review...

**Comment by Borda** [comment]
Todo: ensure back compatibility...

File: .readthedocs.yml [hunk 1]
```diff
@@ -5,9 +5,13 @@
...
 version: 2
 
+# Build documentation in the docs/ directory with Sphinx
+sphinx:
+  configuration: docs/source/conf.py
+
 # Build documentation with MkDocs
-mkdocs:
-  configuration: mkdocs.yml
+#mkdocs:
```
Comment (jeffling) [inline .readthedocs.yml]: Shall we just remove this?

File: .readthedocs.yml [hunk 2]
```diff
@@ -16,4 +20,5 @@ formats: all
 python:
   version: 3.7
   install:
-    - requirements: docs/requirements.txt
\ No newline at end of file
+    #- requirements: requirements.txt
```
Comment (jeffling) [inline .readthedocs.yml]: same with this

File: docs/source/intro.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    model = CoolerNotBERT()
+    
+trainer = Trainer(gpus=4, use_amp=True)
+trainer.fit(model)
+```
+
+Notice a few things about this flow:   
+1. You're writing pure PyTorch... no unnecessary abstractions or new libraries to learn.   
+2. You get free GPU and 16-bit support without writing any of that code in your model.   
+3. You also get all of the capabilities below (without coding or testing yourself).
```
Comment (jeffling) [inline docs/source/intro.md]: 'all the capabilities below' but there is nothing below? does it render differently?

File: pytorch_lightning/callbacks/pt_callbacks.py [hunk 4]
```diff
@@ -355,11 +363,11 @@ def on_epoch_begin(self, epoch, trainer):
...
 
-if __name__ == '__main__':
-    c = EarlyStopping(min_delta=0.9, patience=2, verbose=True)
-    losses = [10, 9, 8, 8, 6, 4.3, 5, 4.4, 2.8, 2.5]
-    for i, loss in enumerate(losses):
-        should_stop = c.on_epoch_end(i, logs={'val_loss': loss})
-        logging.info(loss)
-        if should_stop:
-            break
+# if __name__ == '__main__':
```
Comment (jeffling) [inline pytorch_lightning/callbacks/pt_callbacks.py]: why comment out?

File: tests/test_y_logging.py [hunk 5]
```diff
@@ -78,7 +78,7 @@ def test_mlflow_logger():
     reset_seed()
 
     try:
-        from pytorch_lightning.logging import MLFlowLogger
+        from pytorch_lightning.logging import MlFlowLogger
```
Comment (jeffling) [inline tests/test_y_logging.py]: Is this a real fix?

File: .readthedocs.yml (hunk 1 above)
Comment (Borda) [reply .readthedocs.yml]: we can :)

File: docs/source/intro.md (hunk 3 above)
Comment (Borda) [reply docs/source/intro.md]: it is probably some artefact from copy-paste text to its destinations, but I guess that there will be more of them... suggest to leave the pruning to follow-up PR :]

File: pytorch_lightning/callbacks/pt_callbacks.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/callbacks/pt_callbacks.py]: I do not think that there should be any `__main__` in the package

File: tests/test_y_logging.py (hunk 5 above)
Comment (Borda) [reply tests/test_y_logging.py]: not really, we can swap it back...

File: docs/source/intro.md (hunk 3 above)
Comment (jeffling) [reply docs/source/intro.md]: Agreed!

**Comment by williamFalcon** [comment]
@Borda rebase?

**Comment by Borda** [comment]
yes, working on it right now... :smiley_cat:

**Comment by Borda** [comment]
@williamFalcon rebased... @jeffling one more check, pls

**Comment by Ir1d** [comment]
//offtopic: I suggest we enable netlify or some other tools which support preciewing docs in PRs. As we move from mkdocs to sphinx, it somehow gets harder to read the source directly, and reviewing the built results would be much better. (IMO)

@williamFalcon

**Comment by williamFalcon** [comment]
@Borda tried to build docs but failed.

```
(lightning) user-mbp:docs user$ pip install requirements.txt
Collecting requirements.txt
  ERROR: Could not find a version that satisfies the requirement requirements.txt (from versions: none)
ERROR: No matching distribution found for requirements.txt
(lightning) user-mbp:docs user$ vim requirements.txt
(lightning) user-mbp:docs user$ make clean
(lightning) user-mbp:docs user$ make html

Running Sphinx v2.2.1

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/user/miniconda3/envs/lightning/lib/python3.7/site-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/Users/user/miniconda3/envs/lightning/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 81, in execfile_
    exec(code, _globals)
  File "/Users/user/Developer/pytorch-lightning/docs/source/conf.py", line 23, in <module>
    import pt_lightning_sphinx_theme
ModuleNotFoundError: No module named 'pt_lightning_sphinx_theme'

make: *** [html] Error 2
(lightning) user-mbp:docs user$
(lightning) user-mbp:docs user$ make html
Running Sphinx v2.2.1

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/user/miniconda3/envs/lightning/lib/python3.7/site-packages/sphinx/config.py", line 361, in eval_config_file
    execfile_(filename, namespace)
  File "/Users/user/miniconda3/envs/lightning/lib/python3.7/site-packages/sphinx/util/pycompat.py", line 81, in execfile_
    exec(code, _globals)
  File "/Users/user/Developer/pytorch-lightning/docs/source/conf.py", line 23, in <module>
    import pt_lightning_sphinx_theme
ModuleNotFoundError: No module named 'pt_lightning_sphinx_theme'
```

File: docs/source/_templates/theme_variables.jinja [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{%- set external_urls = {
+  'github': 'https://github.com/williamFalcon/pytorch-lightning',
+  'github_issues': 'https://github.com/williamFalcon/pytorch-lightning/issues',
+  'contributing': 'https://github.com/williamFalcon/pytorch-lightning/blob/master/CONTRIBUTING.md',
+  'docs': 'https://williamfalcon.github.io/pytorch-lightning',
+  'twitter': 'https://twitter.com/pytorch',
```
Comment (williamFalcon) [inline docs/source/_templates/theme_variables.jinja]: pytorch-lightning

**Comment by Borda** [comment]
You need to install docs/requirements.txt

File: pl_examples/basic_examples/lightning_module_template.py [hunk 7]
```diff
@@ -213,17 +213,17 @@ def __dataloader(self, train):
 
         return loader
 
-    @pl.data_loader
+    @data_loader
```
Comment (williamFalcon) [inline pl_examples/basic_examples/lightning_module_template.py]: i prefer pl.data_loader because it's clear where the decorator came from

**Comment by williamFalcon** [comment]
@Borda I followed your instructions:

```
cd docs
pip install requirements.txt   < ------- FAIL HERE
make clean
make html
make latexpdf
```

**Comment by Borda** [comment]
@williamFalcon what is the crash message?
CI builds without any problem https://circleci.com/gh/Borda/pytorch-lightning/298

File: pl_examples/basic_examples/lightning_module_template.py (hunk 7 above)
Comment (Borda) [reply pl_examples/basic_examples/lightning_module_template.py]: I like this import as it is explicit what all function you really need from particular module, but no big deal...

**Comment by williamFalcon** [comment]
the crash message is what i posted above. that’s the result of installing reqs after cd docs

**Comment by awaelchli** [comment]
@williamFalcon in your output above I can see you ran `pip install requirements.txt` without the `-r` option. That's why the imports fail later.

**Comment by williamFalcon** [comment]
right haha. blindly copied that command

**Comment by Borda** [comment]
I have it also tested on RTD, see https://torch-lightning.readthedocs.io/en/docs
I will drop the parallel project just after the merge... :)

**Comment by Borda** [comment]
> right haha. blindly copied that command

copy-paste `you need to install docs/requirements.txt` :-D

**Comment by williamFalcon** [comment]
I'm following your instructions:

```
cd docs
pip install -r requirements.txt (this was wrong and was missing -r. i just fixed it in your instructions
make clean
```

I'm getting these errors:
```
(pytorch-lightning) ➜  pytorch-lightning git:(master) make clean
make: *** No rule to make target `clean'.  Stop.
(pytorch-lightning) ➜  pytorch-lightning git:(master) cd docs
(pytorch-lightning) ➜  docs git:(master) make clean
make: *** No rule to make target `clean'.  Stop.
(pytorch-lightning) ➜  docs git:(master) make html

make: *** No rule to make target `html'.  Stop.
(pytorch-lightning) ➜  docs git:(master)
```

**Comment by Ir1d** [comment]
are you in the folder of correct Makefile?

**Comment by Borda** [comment]
```
pip install -r requirements.txt --user
sudo pip install -r docs/requirements.txt
cd docs
make clean
make html
```
CI builds without any problem https://circleci.com/gh/Borda/pytorch-lightning/298
I have it also tested on RTD, see https://torch-lightning.readthedocs.io/en/docs

**Comment by williamFalcon** [comment]
Yup... still nothing.

```
(pytorch-lightning) ➜  ~ cd ~/Developer/opensource/pytorch-lightning
(pytorch-lightning) ➜  pytorch-lightning git:(master) pip install -r requirements.txt --user

Requirement already satisfied: scikit-learn>=0.20.2 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r requirements.txt (line 1)) (0.20.2)
Requirement already satisfied: tqdm>=4.35.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r requirements.txt (line 2)) (4.35.0)
Requirement already satisfied: numpy>=1.16.4 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r requirements.txt (line 3)) (1.16.4)
Requirement already satisfied: torch>=1.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r requirements.txt (line 4)) (1.3.0)
Requirement already satisfied: torchvision>=0.3.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r requirements.txt (line 5)) (0.4.1a0+d94043a)
Requirement already satisfied: pandas>=0.24 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r requirements.txt (line 6)) (0.24.2)
Requirement already satisfied: test-tube>=0.6.9 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r requirements.txt (line 7)) (0.7.3)
Requirement already satisfied: scipy>=0.13.3 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from scikit-learn>=0.20.2->-r requirements.txt (line 1)) (1.3.1)
Requirement already satisfied: six in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from torchvision>=0.3.0->-r requirements.txt (line 5)) (1.12.0)
Requirement already satisfied: pillow>=4.1.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from torchvision>=0.3.0->-r requirements.txt (line 5)) (6.0.0)
Requirement already satisfied: pytz>=2011k in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from pandas>=0.24->-r requirements.txt (line 6)) (2019.1)
Requirement already satisfied: python-dateutil>=2.5.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from pandas>=0.24->-r requirements.txt (line 6)) (2.8.0)
Requirement already satisfied: imageio>=2.3.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from test-tube>=0.6.9->-r requirements.txt (line 7)) (2.5.0)
Requirement already satisfied: tb-nightly==1.15.0a20190708 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from test-tube>=0.6.9->-r requirements.txt (line 7)) (1.15.0a20190708)
Requirement already satisfied: future in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from test-tube>=0.6.9->-r requirements.txt (line 7)) (0.18.1)
Requirement already satisfied: protobuf>=3.6.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from tb-nightly==1.15.0a20190708->test-tube>=0.6.9->-r requirements.txt (line 7)) (3.8.0)
Requirement already satisfied: absl-py>=0.4 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from tb-nightly==1.15.0a20190708->test-tube>=0.6.9->-r requirements.txt (line 7)) (0.7.1)
Requirement already satisfied: markdown>=2.6.8 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from tb-nightly==1.15.0a20190708->test-tube>=0.6.9->-r requirements.txt (line 7)) (3.1.1)
Requirement already satisfied: grpcio>=1.6.3 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from tb-nightly==1.15.0a20190708->test-tube>=0.6.9->-r requirements.txt (line 7)) (1.21.1)
Requirement already satisfied: wheel>=0.26; python_version >= "3" in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from tb-nightly==1.15.0a20190708->test-tube>=0.6.9->-r requirements.txt (line 7)) (0.33.4)
Requirement already satisfied: werkzeug>=0.11.15 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from tb-nightly==1.15.0a20190708->test-tube>=0.6.9->-r requirements.txt (line 7)) (0.15.4)
Requirement already satisfied: setuptools>=41.0.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from tb-nightly==1.15.0a20190708->test-tube>=0.6.9->-r requirements.txt (line 7)) (41.0.1)
(pytorch-lightning) ➜  pytorch-lightning git:(master) sudo pip install -r docs/requirements.txt

Password:
Sorry, try again.
Password:
Sorry, try again.
Password:
sudo: 2 incorrect password attempts
(pytorch-lightning) ➜  pytorch-lightning git:(master)
(pytorch-lightning) ➜  pytorch-lightning git:(master)
(pytorch-lightning) ➜  pytorch-lightning git:(master)
(pytorch-lightning) ➜  pytorch-lightning git:(master)
(pytorch-lightning) ➜  pytorch-lightning git:(master) pip install -r docs/requirements.txt
Requirement already satisfied: mkdocs-material==4.4.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r docs/requirements.txt (line 1)) (4.4.0)
Requirement already satisfied: mkdocs==1.0.4 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r docs/requirements.txt (line 2)) (1.0.4)
Requirement already satisfied: Pygments>=2.2 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (2.4.1)
Requirement already satisfied: pymdown-extensions>=4.11 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (6.1)
Requirement already satisfied: mkdocs-minify-plugin>=0.2 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (0.2.1)
Requirement already satisfied: Jinja2>=2.7.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (2.10.3)
Requirement already satisfied: PyYAML>=3.10 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (5.1.2)
Requirement already satisfied: Markdown>=2.3.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (3.1.1)
Requirement already satisfied: tornado>=5.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (6.0.3)
Requirement already satisfied: click>=3.3 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (7.0)
Requirement already satisfied: livereload>=2.5.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (2.6.1)
Requirement already satisfied: pep562 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from pymdown-extensions>=4.11->mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (1.0)
Requirement already satisfied: htmlmin>=0.1.4 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-minify-plugin>=0.2->mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (0.1.12)
Requirement already satisfied: jsmin>=2.2.2 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-minify-plugin>=0.2->mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (2.2.2)
Requirement already satisfied: MarkupSafe>=0.23 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from Jinja2>=2.7.1->mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (1.1.1)
Requirement already satisfied: setuptools>=36 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from Markdown>=2.3.1->mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (41.0.1)
Requirement already satisfied: six in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from livereload>=2.5.1->mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (1.12.0)
(pytorch-lightning) ➜  pytorch-lightning git:(master) cd docs

(pytorch-lightning) ➜  docs git:(master) make clean
make: *** No rule to make target `clean'.  Stop.
(pytorch-lightning) ➜  docs git:(master) cd ..
(pytorch-lightning) ➜  pytorch-lightning git:(master) ls
LICENSE                    appveyor.yml               mkdocs.yml                 pytest.log                 setup.cfg                  tox.ini
MANIFEST.in                dist                       pip-wheel-metadata         pytorch_lightning          setup.py                   update.sh
README.md                  docs                       pl_examples                pytorch_lightning.egg-info site
Untitled Folder            lightning_logs             pyproject.toml             requirements.txt           tests
(pytorch-lightning) ➜  pytorch-lightning git:(master) sudo pip install -r docs/requirements.txt

Password:
Sorry, try again.
Password:
Sorry, try again.
Password:
WARNING: The directory '/Users/user/Library/Caches/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
WARNING: The directory '/Users/user/Library/Caches/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Requirement already satisfied: mkdocs-material==4.4.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r docs/requirements.txt (line 1)) (4.4.0)
Requirement already satisfied: mkdocs==1.0.4 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from -r docs/requirements.txt (line 2)) (1.0.4)
Requirement already satisfied: mkdocs-minify-plugin>=0.2 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (0.2.1)
Requirement already satisfied: pymdown-extensions>=4.11 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (6.1)
Requirement already satisfied: Pygments>=2.2 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (2.4.1)
Requirement already satisfied: click>=3.3 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (7.0)
Requirement already satisfied: Jinja2>=2.7.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (2.10.3)
Requirement already satisfied: tornado>=5.0 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (6.0.3)
Requirement already satisfied: PyYAML>=3.10 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (5.1.2)
Requirement already satisfied: livereload>=2.5.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (2.6.1)
Requirement already satisfied: Markdown>=2.3.1 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (3.1.1)
Requirement already satisfied: htmlmin>=0.1.4 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-minify-plugin>=0.2->mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (0.1.12)
Requirement already satisfied: jsmin>=2.2.2 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from mkdocs-minify-plugin>=0.2->mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (2.2.2)
Requirement already satisfied: pep562 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from pymdown-extensions>=4.11->mkdocs-material==4.4.0->-r docs/requirements.txt (line 1)) (1.0)
Requirement already satisfied: MarkupSafe>=0.23 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from Jinja2>=2.7.1->mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (1.1.1)
Requirement already satisfied: six in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from livereload>=2.5.1->mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (1.12.0)
Requirement already satisfied: setuptools>=36 in /Users/user/miniconda3/envs/pytorch-lightning/lib/python3.7/site-packages (from Markdown>=2.3.1->mkdocs==1.0.4->-r docs/requirements.txt (line 2)) (41.0.1)
(pytorch-lightning) ➜  pytorch-lightning git:(master) cd docs
(pytorch-lightning) ➜  docs git:(master) make clean
make: *** No rule to make target `clean'.  Stop.
(pytorch-lightning) ➜  docs git:(master)
```

**Comment by Ir1d** [comment]
hi @williamFalcon are you in master branch or this pr?

**Comment by williamFalcon** [comment]
this pr

**Comment by williamFalcon** [comment]
oh wait, i think it went back to master for some reason. hold on

**Comment by williamFalcon** [comment]
@borda ok, got it to  work.

It looks like a  good starting point but we need to work on the navigation  since it is not  intuitive at all right now nor obvious where to find things.

Can we deploy this behind a dev endpoint to see what it looks like right now?

Things that need to be fixed:
- simple, intuitive menu on the left
- logo

This is a good start, so I'm just going to highlight a few places we should focus on to get it to a good release point.

I assume the people who look in places like this are people contributing to lightning, but we need to show it in a more user friendly way.
image

This menu should have a Trainer and LightningModule root folder, then a kind of similar structure to the docs we have right now.

Again, this is an amazing v1, so great job! Let's now move to make this more user friendly before we fully switch over to these docs.
image

**Comment by williamFalcon** [comment]
@Borda I approve to merge this right now (again phenomenal job!).

@neggert @jeffling @Ir1d  you guys good to merge as well? this is a big PR so want to get multiple eyes/approvals.

**Comment by Ir1d** [comment]
Talking about dev endpoint.. I'm suggesting netlify again.
Meanwhile it seems @Borda has built this PR on readthedocs

**Comment by Borda** [comment]
> Can we deploy this behind a dev endpoint to see what it looks like right now?

Not sure about his, it is a bit bond my experience so far...

> Things that need to be fixed:
> * simple, intuitive menu on the left
> * logo

What do you mean by logo, size?

> This is a good start, so I'm just going to highlight a few places we should focus on to get it to a good release point.

for sure, I just wanted to keep the PR straight forward and draw the backbone so since this PR the work can be simpler divided (parallelize) into smaller PRs
btw, also each rebases after the master change and solving collision is a bit painful...

**Comment by Borda** [comment]
@Ir1d I believe that the actual docs is also powered by RTD according to the [config](https://github.com/williamFalcon/pytorch-lightning/blob/master/.readthedocs.yml) in root

**Comment by Ir1d** [comment]
Oh I didn't notice this config before. @Borda Thx
Netlify is really good for deploy previews for docs, saves a lot of time for cloning and builing. https://www.netlify.com/tags/deploy-previews/

**Comment by Borda** [comment]
@Ir1d RTD allows you to have a separate version for each branch even for each tag/release

**Comment by williamFalcon** [comment]
@Borda for the record we use pl not pt

```python
import pytorch_lightning as pl
```

**Comment by Borda** [comment]
> @Borda for the record we use pl not pt

typo :angel:

**Comment by Borda** [comment]
FYI, this is the fork we are right now using for docs - https://github.com/Borda/lightning_sphinx_theme

**Comment by jeffling** [comment]
Nice job bringing this in!

**Pull request merged**